app/models/role.rb

```diff
@@ -146,31 +146,78 @@ def add_permissions(permissions, options = {})
     permissions = Array(permissions)
     search = options.delete(:search)
 
-    collection = Permission.where(:name => permissions).all
-    raise ::Foreman::PermissionMissingException.new(N_('some permissions were not found')) if collection.size != permissions.size
+    collection = permission_records permissions
 
     current_filters = self.filters
     collection.group_by(&:resource_type).each do |resource_type, grouped_permissions|
-      filter = find_filter resource_type, current_filters, search
+      filter = filter_for_permission_add resource_type, current_filters, search
 
       grouped_permissions.each do |permission|
+        next if filter.permissions.include?(permission)
         filtering = filter.filterings.build
         filtering.filter = filter
         filtering.permission = permission
       end
     end
   end
 
+  def remove_permissions(permissions)
+    permissions = Array(permissions)
+    collection = permission_records permissions
+    current_filters = self.filters
+    find_permissions_to_remove collection, current_filters
+  end
+
+  def find_permissions_to_remove(collection, current_filters)
```

Comment:
I split this because code climate was complaining about method complexity, I'll put it back together.